internal/provider/meta.go

```diff
@@ -216,40 +217,93 @@ func NewProviderMeta(d *schema.ResourceData) (interface{}, error) {
 
 	MaxHTTPRetriesCCC = d.Get("max_retries_ccc").(int)
 
-	// Try and get the token from the config or token helper
-	token, err := GetToken(d)
-	if err != nil {
-		return nil, err
-	}
+	// Set the namespace to the requested namespace, if provided
+	namespace := d.Get(consts.FieldNamespace).(string)
 
 	authLogin, err := GetAuthLogin(d)
 	if err != nil {
 		return nil, err
 	}
 
+	var token string
 	if authLogin != nil {
-		client.SetNamespace(authLogin.Namespace())
-		secret, err := authLogin.Login(client)
+		// the clone is only used to auth to Vault
+		clone, err := client.Clone()
+		if err != nil {
+			return nil, err
+		}
+
+		if authLogin.Namespace() != "" {
+			// the namespace configured on the auth_login takes precedence over the provider's
+			// for authentication only.
+			clone.SetNamespace(authLogin.Namespace())
+		} else if namespace != "" {
+			// authenticate to the engine in the provider's namespace
+			clone.SetNamespace(namespace)
+		}
+
+		secret, err := authLogin.Login(clone)
 		if err != nil {
 			return nil, err
 		}
 
 		token = secret.Auth.ClientToken
+	} else {
+		// try and get the token from the config or token helper
+		token, err = GetToken(d)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	if token != "" {
 		client.SetToken(token)
 	}
+
 	if client.Token() == "" {
-		return nil, errors.New("no vault token found")
+		return nil, errors.New("no vault token set on Client")
 	}
 
-	skipChildToken := d.Get("skip_child_token").(bool)
-	if !skipChildToken {
-		err := setChildToken(d, client)
+	tokenInfo, err := client.Auth().Token().LookupSelf()
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup token, err=%w", err)
+	}
+
+	var tokenNamespace string
+	if v, ok := tokenInfo.Data[consts.FieldNamespacePath]; ok {
+		tokenNamespace = strings.Trim(v.(string), "/")
+	}
+
+	if !d.Get(consts.FieldSkipChildToken).(bool) {
+		// a child token is always created in the namespace of the parent token.
+		token, err = createChildToken(d, client, tokenNamespace)
 		if err != nil {
 			return nil, err
 		}
+
+		client.SetToken(token)
+	}
+
+	if namespace == "" && tokenNamespace != "" {
```

Comment:
How do I specify that I want provisioning to take place in the root namespace if my token is using a different namespace?

Reply:
You would need to set the namespace here: 
https://registry.terraform.io/providers/hashicorp/vault/latest/docs#namespace

Reply:
The root namespace is named `""`, how do you distinguish that from namespace not set?

Reply:
I think that correctly handling the default namespace is out of scope for this PR. We can fix that in the future, since it could introduce more complications for the user.